Add HIP for validator challenges (#359)
helium/HIP: xxxx-validator-challenges.md

connectivity failures. Hotspots can use Validators to learn about chain updates,
including if they're being challenged.

# Stakeholders
[stakeholders]: #stakeholders

* Who is affected by this HIP?

Users running Hotspots or Validators.

* How are we soliciting feedback on this HIP from these stakeholders? Note that
  they may not be watching the HIPs repository or even aren't directly active in
  the Helium Community Slack channels.

# Detailed Explanation
[detailed-explanation]: #detailed-explanation

## Description of Current Implementation

First it is probably useful to recapitulate the current challenge process. First
refer to the Detailed Explanation section of the [H3Dex Targeting HIP](link when
merged) which describes how a challenge is constructed and submitted to the
chain.

Once the PoC request has appeared on chain and the Challengee has been selected
the Hotspot attempts to connect to it over p2p (failure point #1). If successful
it sends the challengee the data to transmit. The Challengee then consults its
ledger to see if the PoC is active (failure point #2 if the Challengee is not
synced). The Challengee then transmits the packet and sends a receipt to the

> Challengeer over the same connection the Challenger dialed on (this avoids

Reviewer: ```suggestion
Challenger over the same connection the Challenger dialled on (this avoids
```

Reviewer 2: resolved in #361

expired).

So, as can be seen from above there are a number of failure points in this
process. These failures are all related to the speed/size of the chain and the
size of the p2p network. When the network was smaller this was much less of an
issue, but with over 500,000 Hotspots on the network now and larger blocks these
failure points are becoming more and more common.

An important aspect of this system is that the actual challenge Entropy is kept
secret until the receipt transaction is published. This requires Hotspots to
observe the actual data from the Challenger to correctly respond to the
challenge and prevents poisoning attacks. However Hotspots can verify the
challenge information against the chain information once they see it.

### Light Gateways & GRPC

Light gateways will maintain a durable GRPC connection to a validator.  The
target validator could be random or specified.  Via this durable connection,
light gateways will be streamed chain related data, events and notifications.

Light gateways can also connect to additional validators as required.  Such
connections will be ephemeral in nature and typically for unary requests.

> All message types sent to a light gateway from a validator will be wrapped in a
> top level msg of the type 'gateway_resp_v1' ( see proto link below ).  This msg
> type includes a signed payload of the msg and in addition the block height at
> which the msg contents were valid.  Together these two fields serve as an
> attestation on behalf of the sending validator.

Reviewer: ```suggestion
All message types sent to a light gateway from a validator will be wrapped in a
top level msg of the type 'gateway_resp_v1'.  This msg
type includes a number of fields which together serve as an attestation on behalf of the sending validator.  The attestation data includes:
- A signed payload of the msg
- Block age
- Block Height
- Block Time ```

Reviewer 2: Addressed in #361

ChallengeRate challenges per block.

The set of agreed on public keys hashes will then be deterministically truncated
to ChallengeRate and form part of the block metadata. 

Upon handling a block, each CG member will inspect the public keys in the block
and identify any of their own keys and for each initiate a new PoC.   In running
the PoC the public key hash will be used along with the block hash to generate
entropy via which a H3 region will be selected.  Entropy from a combination of
the associated private key hash and the block hash will then be used to identify
the target within the region and subsequently the challenge itself will be
generated.  The region, target and challenge will be persisted to disk.

### Validators & Light Gateway Communication

All validators, whether they are in consensus or not, as part of handling each
block will pull the public key hashes from the block metadata and together with
the block hash generate the same entropy as the CG members to identify the
target region.  The validators will then pull a list of all gateways within that
region and send each gateway in that region connected to them a notification msg

> of type 'poc_challenge_notification_resp_v1' informing them of a challenge

Reviewer: ```suggestion
 informing them of a challenge
```

Reviewer 2: Resolved in #361

> within their region. The payload  provides the necessary routing data ( pubkey

Reviewer: ```suggestion
within their region. The notification  provides the necessary routing data ( pubkey
```

Reviewer 2: resolved in #361

and IP) to enable the light gateway to connect to the challenging validator and
in addition includes the onion key hash of the challenge.

> All gateways upon receipt of a challenge notification will respond with a unary
> GRPC request of type 'gateway_poc_check_challenge_target_req_v1' to the
> challenging CG member to check if they are the target.

Reviewer: ```suggestion
All gateways upon receipt of a challenge notification will send a request over GRPC to the
challenging Validator to check if they are the target.
```

Reviewer 2: Resolved in #361

responsible for block production, Challenges/Witnesses, etc. With the switch to
Validators we moved beyond that model for block production, but we still have
significant computational overhead and complexity on Hotspots as a result of the
old design and constraints of Proof-of-Coverage.

This complexity has become a significant pain point, Hotspots must now keep up
with a blockchain that is produced on significantly more powerful hardware and
they must contend with an enormous peer-to-peer (p2p) network to route
Challenges and witness reports. The global chip shortage have also made it
harder to source capable hardware for building Hotspots that can meet these
requirements.

To address these issues the core developers have been working on design and
implementation of an alternative PoC Challenge mechanism we call Validator
Challenges. In brief, Validator Challenges move the role of generating
Challenges to the Consensus Group. This not only allows us to free Hotspots from
the burden of following the blockchain but it also moves the entities initiating
Challenges to machines with much more stable and predictable networking which

> reduces the likihood of connectivity failures. Hotspots can become clients of

Reviewer 3: ```suggestion
reduces the likelihood of connectivity failures. Hotspots can become clients of
```

Validators to learn about blockchain updates in general, whether or not they're
currently being challenged, and where to deliver Witness receipts.

# Stakeholders
[stakeholders]: #stakeholders

Hotspot and Validator owners/operators.

# Detailed Explanation
[detailed-explanation]: #detailed-explanation

## Description of Current Implementation

It is useful to recapitulate the current challenge process. First refer to the
Detailed Explanation section of [HIP53: H3Dex Targeting][hip53] which describes
how a challenge is constructed and submitted to the blockchain.

1. Once the PoC Request has appeared on the blockchain and the Challengee has
   been selected, the Hotspot attempts to connect to it over p2p (potential
   failure point). If successful it sends the challengee the data to transmit.
2. The Challengee then consults its local blockchain ledger to see if the PoC is
   active. This is another potential failure point if the Challengee is not
   synced.
3. The Challengee then transmits the packet and sends a receipt to the

>    Challenger over the same connection from the Challenger. Tthis avoids another

Reviewer 3: ```suggestion
   Challenger over the same connection from the Challenger. This is another
```

   potential p2p failure.

> 4. Any Hotspots observing the challenge packet then consults their local
>    blockchain ledger to see if they can resolve the challenge to a Challenger.
>    This is another potential failure point if the witness is not in sync with
>    the blockchain.
> 5. If the challenge can be resolved to a Challenger then the Witness attempts to
>    dial that Challenger over p2p, another potential failure point.

Reviewer 2: ```suggestion
4. Any Hotspots observing the challenge packet then consults their local
   blockchain ledger to see if they can find the Challenger based on the packet
   they received and Challenge data on the blockchain. This is another potential
   failure point if the witness is not in sync with the blockchain.
5. If the Challenger is found, the Witnessing Hotspot attempts to dial them over
   p2p, another potential failure point.
```

6. Once the Challenger has decided enough blocks have passed it submits the
   `poc_receipt` transaction to the blockchain to report on the information

>    gathered (if any. The final, potential failure point is if the Challenger is

Reviewer 3: ```suggestion
   gathered (if any). The final, potential failure point is if the Challenger is
```

   lagging behind the blockchain and takes so long to submit the receipt that
   the PoC has expired.

So, as can be seen from above there are a number of potential failure points in
this process. These failures are all related to the speed/size of the blockchain
and the size of the p2p network. When the network was smaller this was much less
of an issue, but with over 500,000 Hotspots on the network now and larger
blocks, these failures are becoming more and more common.

An important aspect of this system is that the actual challenge entropy is kept
secret until the receipt transaction is published. This requires Hotspots to
observe the actual data from the Challenger to correctly respond to the
challenge and prevents poisoning attacks. However Hotspots can verify the
challenge information against the blockchain once they see it.

> ## Light Hotspots & GRPC

Reviewer 3: ```suggestion
## Light Hotspots & gRPC
```

> Light Hotspots will maintain a durable GRPC connection to one (or many)

Reviewer 3: ```suggestion
Light Hotspots will maintain a durable GRPC (Google Remote Procedure Call) connection to one (or many)
```

Validators.  The target Validator(s) could be random or specified.  Via the
first durable connection, Light Hotspots will be streamed chain related data,
events and notifications. Additional connections can be considered ephermeral

> and used for lookup requests if neded.

Reviewer 3: ```suggestion
and used for lookup requests if needed.
```

block proposals, whilst the private key will be saved to local state on the generating
Validator. We propose a fixed `poc_challenge_rate` parameter to be added to the
chain that defines the target number of Challenges per block. Each Consensus
Group member will propose a certain number of keys in order to ensure that we
will meet this target. Assuming there are a minimum of `2f+1` nodes
participating in a block, we are able to reach `poc_challenge_rate` if each
Validator in the Group generates enough Challenges to fulfill this formula:

![\frac{\text{PoC Challenge Rate}}{\frac{N-1}{3}* 2}][poc-challenge-rate]

The set of agreed on public keys hashes will then be deterministically truncated
(in case there are more than `2f+1` participating to `poc_challenge_rate` and
form part of the block metadata. The advantage of a fixed `poc_challenge_rate`
is that regardless of Hotspot growth, this value can remain fixed unless changed
through governance and can be adjusted based on the ability for Validators to
serve the capacity needs of the network. This avoids the need for periodic
changes to `poc_challenge_interval` as we do today in order to reduce load on
the network.

> Upon handling a block, each CG member will inspect the public keys in the block,
> identify any of their own keys, and for each, initiate a new PoC. The public key
> hash will be used with the block hash to generate entropy to generate an H3
> region for the Challenge.  Entropy from a combination of the associated private
> key hash and the block hash will then be used to identify the target within the
> region to generate the Challenge itself. The region, challenge, and target will
> be persisted locally on the Challenger Validator.

Reviewer 2: ```suggestion
Upon handling a block, each Consensus Group member will inspect the public keys
in the block, identify any of their own keys, and for each, initiate a new PoC.
The public key hash will be used with the block hash to generate entropy to
generate an H3 region for the Challenge.  Entropy from a combination of the
associated private key hash and the block hash will then be used to identify the
target within the region to generate the Challenge itself. The region,
challenge, and target will be persisted locally on the Challenger Validator.
```

data payloads exchanged between the Light Hotspots, the Validators, and the
Consensus Group will include attestation data. This can can enable a slashing
mechanism.  It does not however propose the slashing implementation itself, that
could be proposed in a future HIP.

This plan also allows for an establishment of tracking which Validators that
Hotspots are using for their source of blockchain information but it does not
propose what to use that information for, like affinity or delegation. We leave
this for a future proposal as well.

This plan, finally, allows for a change to how Challenges work and would allow
for multiple challengees to be involved in transmitting packets. This could
reduce the scope for collusion between a Challengee and Witnesses and also
obtain better positional information as a result of the Challenge. This has been
described in other forums as "Region" or "Multi-Hotspot" Challenges. This is out
of scope for this HIP but we imagine future work in this direction could be
possible.

# Deployment Impact
[deployment-impact]: #deployment-impact

Overall users should see more reliable Challenge behaviour, less data use on the

> Hotspot (which makes cellular/satellite backhaul hotspots more feasible) and

Reviewer 3: ```suggestion
Hotspot (which makes cellular/satellite backhaul Hotspots more feasible) and
```
